### PR TITLE
Hoofstuk 1/2/begin 3 opgaves van boek 

### PR DESCRIPTION
**Pull request by jobvancreij**

### CODE REVIEW

File: fps/src/main/scala/chapter01/simpleExamp.scala [hunk 1]
```diff
@@ -1,64 +1,71 @@
...
+      val (z,t) = purchases.unzip
+      (z,t.reduce((x, y) => x.combine(y)))
+
+
+ }
+}
+
+
+
+object simpleExamp {
```
Comment (nmcb) [inline fps/src/main/scala/chapter01/simpleExamp.scala:54]: prefer application extension, ie.

```
object simpleExamp extends App:
  ...
  println(...)
```

File: fps/src/main/scala/chapter02/exercise2_1.scala [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package chapter02
+
+
+
+object exercise2_1 extends App {
+
+  //2.1
+  def calcFib(n: Int): Int =
```
Comment (nmcb) [inline fps/src/main/scala/chapter02/exercise2_1.scala:8]: try to implement fib with library functions also, ie. non recursively

File: fps/src/main/scala/chapter02/exercise2_2.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    b == a + a
+  }
+
+  def isSorted[A](as: Array[A], ordered: (A,A) => Boolean): Boolean = {
+    // Make list with tuple of all combinations
+    // check the comparison function per combination
+    // check if all are true
+    as.dropRight(1)
+      .zip(as.tail)
+      .map((a,b) => ordered(a,b))
```
Comment (nmcb) [inline fps/src/main/scala/chapter02/exercise2_2.scala:16]: can use ie. `_ ordered _`

File: fps/src/main/scala/chapter02/exercise2_2.scala [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      .forall(x => x)
```
Comment (nmcb) [inline fps/src/main/scala/chapter02/exercise2_2.scala:17]: prefer use of `identity` for `x => x` implementations

File: fps/src/main/scala/chapter02/exercise2_2.scala [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+  def simpleSort(a: String, b: String): Boolean = {
+    b == a + a
+  }
+
+  def isSorted[A](as: Array[A], ordered: (A,A) => Boolean): Boolean = {
+    // Make list with tuple of all combinations
+    // check the comparison function per combination
+    // check if all are true
+    as.dropRight(1)
```
Comment (nmcb) [inline fps/src/main/scala/chapter02/exercise2_2.scala:14]: `as.zip(as.tail)` also works - ie. `zip` returns a zipped list with the same length as the shortest list, ie. dropping the elements of the longer list, tail side

File: fps/src/main/scala/chapter03/exercise3_1.scala [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package chapter03
+
+import chapter03.exercise3_1.List.sum
+
+//(3, 14, 12,9,101,42
+object exercise3_1 extends App {
+  sealed trait List[+A]
```
Comment (nmcb) [inline fps/src/main/scala/chapter03/exercise3_1.scala:7]: your first ADT - w00t!

File: fps/src/main/scala/chapter03/exercise3_1.scala [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+    def product(ds: List[Double]): Double = ds match {
+      case Nil => 1.0
+      case Cons(0.0, _) => 0.0
+      case Cons(x, xs) => x * product(xs)
+    }
+
+    def apply[A](as: A*): List[A] =
+      if (as.isEmpty) Nil
+      else Cons(as.head, apply(as.tail: _*))
```
Comment (nmcb) [inline fps/src/main/scala/chapter03/exercise3_1.scala:25]: fyi - scala 3 syntax calling apply is `apply(as.tail*)`, ie. mirroring the `as: A*` type definition syntax

File: fps/src/main/scala/chapter03/exercise3_2.scala [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    def getTail[A](ds: List[A]): List[A] = {
+      ds match {
+        case Nil=> Nil
+        case Cons(x,y) => y
+      }
+    }
+
+    def apply[A](as: A*): List[A] =
+      if (as.isEmpty) Nil
+      else Cons(as.head, apply(as.tail: _*))
```
Comment (nmcb) [inline fps/src/main/scala/chapter03/exercise3_2.scala:31]: sanity check question - is apply tail recursive ???

File: fps/src/main/scala/chapter03/exercise3_3.scala [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+      case Nil => 0
+      case Cons(x, xs) => x + sum(xs)
+    }
+
+    def product(ds: List[Double]): Double = ds match {
+      case Nil => 1.0
+      case Cons(0.0, _) => 0.0
+      case Cons(x, xs) => x * product(xs)
+    }
+    def setHead[A](ds: List[A], newVal: A): List[A] = {
```
Comment (nmcb) [inline fps/src/main/scala/chapter03/exercise3_3.scala:22]: fyi - doesn't really "sets" the head, but returns a _new_ list with the head concatenated which is why this function normally is named `concat`

File: fps/src/main/scala/chapter03/exercise3_4.scala [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  object List {
+    def sum(ints: List[Int]): Int = ints match {
+      case Nil => 0
+      case Cons(x, xs) => x + sum(xs)
+    }
+
+    def product(ds: List[Double]): Double = ds match {
+      case Nil => 1.0
+      case Cons(0.0, _) => 0.0
+      case Cons(x, xs) => x * product(xs)
```
Comment (nmcb) [inline fps/src/main/scala/chapter03/exercise3_4.scala:20]: note - monoidal :))

File: fps/src/main/scala/chapter03/exercise3_4.scala [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    def dropN[A](ds: List[A], n: Int): List[A] = {
+      if (n == 0 ){ throw Exception("cannot be zero")}
```
Comment (nmcb) [inline fps/src/main/scala/chapter03/exercise3_4.scala:23]: why can we not drop zero elements from a list ???

File: fps/src/main/scala/chapter03/exercise3_5.scala [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+      case Nil => 0
+      case Cons(x, xs) => x + sum(xs)
+    }
+
+    def product(ds: List[Double]): Double = ds match {
+      case Nil => 1.0
+      case Cons(0.0, _) => 0.0
+      case Cons(x, xs) => x * product(xs)
+    }
+    def dropWhile[A](ds: List[A], f: A => Boolean): List[A] = {
```
Comment (nmcb) [inline fps/src/main/scala/chapter03/exercise3_5.scala:22]: trick question - is this method tail recursive ???

**Review by nmcb — CHANGES_REQUESTED** [review]
nice! minor comments and remarks for educational purposes.

File: fps/src/main/scala/chapter01/simpleExamp.scala [hunk 13]
```diff
@@ -1,64 +1,71 @@
...
-//
-//  }
-//}
-//
-//
-//
-//
-//
+
+class Coffee {
```
[the remark this answers is not in the document]
Comment (jobvancreij) [reply fps/src/main/scala/chapter01/simpleExamp.scala]: #9

File: fps/src/main/scala/chapter01/simpleExamp.scala [hunk 14]
```diff
@@ -1,64 +1,71 @@
...
+  val price = 2.50
+}
+
+ class CreditCard {
+  var valOnCard: Double = 100.00 //have to change it but cannot access vars
```
[the remark this answers is not in the document]
Comment (jobvancreij) [reply fps/src/main/scala/chapter01/simpleExamp.scala]: #9

File: fps/src/main/scala/chapter01/simpleExamp.scala [hunk 15]
```diff
@@ -1,64 +1,71 @@
...
+    if (card == other.card)
+      ChargeCard(card, amount + other.amount)
+    else
+      throw new Exception("different cards")
+  }
+
+}
+class Cafe {
+  def buyCoffee(card: CreditCard): (Coffee,  ChargeCard) = {
+    val cup = new Coffee()
```
[the remark this answers is not in the document]
Comment (jobvancreij) [reply fps/src/main/scala/chapter01/simpleExamp.scala:35]: #9

File: fps/src/main/scala/chapter01/simpleExamp.scala [hunk 16]
```diff
@@ -1,64 +1,71 @@
...
+  def charge(amount: Double): Unit = {
+    valOnCard = valOnCard - amount // fix notation
+  }
+
+}
+
+case class ChargeCard(card: CreditCard, amount: Double) { // why case class?
+  def combine(other: ChargeCard): ChargeCard =  {
+    if (card == other.card)
+      ChargeCard(card, amount + other.amount)
```
[the remark this answers is not in the document]
Comment (jobvancreij) [reply fps/src/main/scala/chapter01/simpleExamp.scala]: #9

File: fps/src/main/scala/chapter01/simpleExamp.scala [hunk 17]
```diff
@@ -1,64 +1,71 @@
...
+    else
+      throw new Exception("different cards")
```
[the remark this answers is not in the document]
Comment (jobvancreij) [reply fps/src/main/scala/chapter01/simpleExamp.scala:29]: #9

File: fps/src/main/scala/chapter01/simpleExamp.scala (hunk 13 above)
[the remark this answers is not in the document]
Comment (jobvancreij) [reply fps/src/main/scala/chapter01/simpleExamp.scala]: #9

File: fps/src/main/scala/chapter01/simpleExamp.scala [hunk 18]
```diff
@@ -1,64 +1,71 @@
...
+class Cafe {
+  def buyCoffee(card: CreditCard): (Coffee,  ChargeCard) = {
+    val cup = new Coffee()
+    (cup, ChargeCard(card: CreditCard, cup.price))
+  }
+
+    def buyCoffees(card: CreditCard, n: Int): (List[Coffee], ChargeCard) = {
+      val purchases: List[(Coffee, ChargeCard)] = List.fill(n)(buyCoffee(card))
+      //val (Coffees: List[Coffee], Charges: List[ChargeCard]) = purchases.unzip
+      val (z,t) = purchases.unzip
```
[the remark this answers is not in the document]
Comment (jobvancreij) [reply fps/src/main/scala/chapter01/simpleExamp.scala]: #9

File: fps/src/main/scala/chapter01/simpleExamp.scala [hunk 19]
```diff
@@ -1,64 +1,71 @@
...
+      (z,t.reduce((x, y) => x.combine(y)))
```
[the remark this answers is not in the document]
Comment (jobvancreij) [reply fps/src/main/scala/chapter01/simpleExamp.scala]: #9

File: fps/src/main/scala/chapter01/simpleExamp.scala (hunk 1 above)
Comment (jobvancreij) [reply fps/src/main/scala/chapter01/simpleExamp.scala:49]: #9

File: fps/src/main/scala/chapter03/exercise3_2.scala (hunk 8 above)
Comment (jobvancreij) [reply fps/src/main/scala/chapter03/exercise3_2.scala:31]: At the moment not. I cannot get it recursive bc . Could we look at this together?

File: fps/src/main/scala/chapter03/exercise3_5.scala (hunk 12 above)
Comment (jobvancreij) [reply fps/src/main/scala/chapter03/exercise3_5.scala:22]: Can we also look at this, currenlty not

**Review by nmcb — APPROVED** [review]
nice - lgtm - so lets move on - will pick up other syntactical prefs in subsequent prs.

File: fps/src/main/scala/chapter03/exercise3_5.scala (hunk 12 above)
Comment (nmcb) [reply fps/src/main/scala/chapter03/exercise3_5.scala:22]: will address in on-on-ones

File: fps/src/main/scala/chapter02/exercise2_1.scala (hunk 2 above)
Comment (nmcb) [reply fps/src/main/scala/chapter02/exercise2_1.scala:8]: we will address this point in the one-on-ones

File: fps/src/main/scala/chapter02/exercise2_2.scala (hunk 3 above)
Comment (nmcb) [reply fps/src/main/scala/chapter02/exercise2_2.scala:16]: fyi

File: fps/src/main/scala/chapter02/exercise2_2.scala (hunk 4 above)
Comment (nmcb) [reply fps/src/main/scala/chapter02/exercise2_2.scala:17]: fyi

File: fps/src/main/scala/chapter03/exercise3_1.scala (hunk 6 above)
Comment (nmcb) [reply fps/src/main/scala/chapter03/exercise3_1.scala:7]: just happy for you

File: fps/src/main/scala/chapter03/exercise3_1.scala (hunk 7 above)
Comment (nmcb) [reply fps/src/main/scala/chapter03/exercise3_1.scala:25]: fyi

File: fps/src/main/scala/chapter03/exercise3_2.scala (hunk 8 above)
Comment (nmcb) [reply fps/src/main/scala/chapter03/exercise3_2.scala:31]: will address this in the one-on-ones

File: fps/src/main/scala/chapter03/exercise3_4.scala (hunk 11 above)
Comment (nmcb) [reply fps/src/main/scala/chapter03/exercise3_4.scala:23]: will address this in the one-on-ones

**Pull request merged**